tests/acceptance/30_generic_methods/file_ensure_key_value.cf

```diff
@@ -0,0 +1,154 @@
+Key=Value";
+
+    "ref_text2"   string=> "# File with lhs=rhs text
+Key=Value
+AddedKey=AddedValue";
+
+    "key2"        string => "AddedKey";
+    "value2"      string => "AddedValue";
+    "separator2"  string => "=";
+
+# Third test: ensure that if key-value is already correct, nothing is done
+    "base_text3"  string => "# File with lhs rhs text
+Key Value";
+
+    "ref_text3"   string => "# File with lhs rhs text
+Key Value";
+
+    "key3"        string => "Key";
+    "value3"      string => "Value";
+    "separator3"  string => " ";
+
+  commands:
+# Initialize test and reference files
+      "/bin/echo"
+        args    => "\"${ref_text1}\" > \"${ref_file1}\"",
+        contain => in_shell;
+      "/bin/echo"
+        args    => "\"${base_text1}\" > \"${file1}\"",
+        contain => in_shell;
+
+      "/bin/echo"
+        args    => "\"${ref_text2}\" > \"${ref_file2}\"",
+        contain => in_shell;
+      "/bin/echo"
+        args    => "\"${base_text2}\" > \"${file2}\"",
+        contain => in_shell;
+
+      "/bin/echo"
+        args    => "\"${ref_text3}\" > \"${ref_file3}\"",
+        contain => in_shell;
+      "/bin/echo"
+        args    => "\"${base_text3}\" > \"${file3}\"",
+        contain => in_shell;
+
+
+}
+
+#######################################################
+
+bundle agent test
+{
+  methods:
+    "modified key" usebundle => file_ensure_key_value("${init.file1}", "${init.key1}", "${init.value1}", "${init.separator1}");
+    "added key"    usebundle => file_ensure_key_value("${init.file2}", "${init.key2}", "${init.value2}", "${init.separator2}");
+    "success key"  usebundle => file_ensure_key_value("${init.file3}", "${init.key3}", "${init.value3}", "${init.separator3}");
+
+}
+
+#######################################################
+
+bundle agent check
+{
+  vars:
+      "file_diff_test1" string => "/usr/bin/diff \"${init.ref_file1}\" \"${init.file1}\"";
+      "file_diff_test2" string => "/usr/bin/diff \"${init.ref_file2}\" \"${init.file2}\"";
+      "file_diff_test3" string => "/usr/bin/diff \"${init.ref_file3}\" \"${init.file3}\"";
+
+  classes:
+    "file1_correct" 
+       expression => returnszero("${file_diff_test1}", "noshell"),
+       ifvarclass => canonify("file_ensure_key_value_${init.file_canon1}_reached");
+    "file2_correct"
+       expression => returnszero("${file_diff_test2}", "noshell"),
+       ifvarclass => canonify("file_ensure_key_value_${init.file_canon2}_reached");
+    "file3_correct"
+       expression => returnszero("${file_diff_test3}", "noshell"),
+       ifvarclass => canonify("file_ensure_key_value_${init.file_canon3}_reached");
+
+
+    "ok_test1"     expression => "file1_correct.file_ensure_key_value_${init.file_canon1}_repaired.!file_ensure_key_value_${init.file_canon1}_not_ok";
+    "ok_test2"     expression => "file2_correct.file_ensure_key_value_${init.file_canon2}_repaired.!file_ensure_key_value_${init.file_canon2}_not_ok";
```

Comment:
That is *exactly* what I was concerned about. What you're saying right there is CFEngine pseudo-logic, and not at all logic to the rest of the human species.

In ncf, we aim to have simple, clean methods: file_ensure_key_value should be repaired in the file needed changes, and kept if the file was already OK, nothing else.

Can you look into a workaround to avoid this confusing and nasty kept class? If it comes from the "create => true" part of the promise, maybe just separate this into two promises, and report only on the second? If not, how about a checksum before and after the file edit promise, and if the checksum differs, then force the repaired class, if not force the kept class?